In [3]:
import pandas as pd
from faker import Faker
from datetime import datetime, timezone, timedelta

In [4]:
fake = Faker()
datetimes = []

In [5]:
datetimes = []
for _ in range(100):  # Adjust the number (100 in this case) to generate the desired number of datetimes
    dt = fake.date_time_this_year()
    if dt.year == 2023:
        datetimes.append(dt)
datetimes

[datetime.datetime(2023, 2, 15, 9, 42, 29),
 datetime.datetime(2023, 5, 14, 4, 36, 50),
 datetime.datetime(2023, 2, 21, 4, 18, 36),
 datetime.datetime(2023, 3, 29, 22, 8, 8),
 datetime.datetime(2023, 3, 20, 7, 54, 18),
 datetime.datetime(2023, 6, 6, 12, 17, 58),
 datetime.datetime(2023, 5, 6, 17, 52, 56),
 datetime.datetime(2023, 4, 26, 2, 26, 19),
 datetime.datetime(2023, 1, 25, 21, 59, 19),
 datetime.datetime(2023, 5, 5, 11, 58, 41),
 datetime.datetime(2023, 5, 13, 21, 42, 3),
 datetime.datetime(2023, 1, 27, 21, 52, 30),
 datetime.datetime(2023, 1, 20, 20, 34, 40),
 datetime.datetime(2023, 5, 13, 9, 45, 30),
 datetime.datetime(2023, 1, 17, 7, 39, 32),
 datetime.datetime(2023, 6, 22, 12, 58, 25),
 datetime.datetime(2023, 3, 30, 14, 41, 14),
 datetime.datetime(2023, 5, 23, 7, 5, 26),
 datetime.datetime(2023, 3, 7, 14, 9, 12),
 datetime.datetime(2023, 4, 7, 15, 2, 19),
 datetime.datetime(2023, 4, 30, 18, 32, 47),
 datetime.datetime(2023, 4, 14, 11, 38, 22),
 datetime.datetime(2023, 6, 1

In [6]:
tz_string = datetime.now().astimezone()
tz_string

datetime.datetime(2023, 6, 23, 11, 25, 15, 265286, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000), 'EDT'))

In [7]:
    # dt = fake.date_time_between_dates(
    #     datetime_start=date_start,
    #     datetime_end=date_end,
    #     tzinfo=tz
    #     )

In [8]:
datetimes = []
date_start =  datetime(2023, 6, 23, 11, 0, 1)
date_end =  datetime(2023, 6, 23, 18, 0, 0)
tz = timezone(timedelta(days=-1, seconds=72000), 'EDT')
n = 10000

for _ in range(n):

    dt = fake.future_datetime(
        end_date="+2h",
        tzinfo=tz
        )
    datetimes.append(dt)
datetimes

[datetime.datetime(2023, 6, 23, 11, 50, 13, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000), 'EDT')),
 datetime.datetime(2023, 6, 23, 11, 48, 37, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000), 'EDT')),
 datetime.datetime(2023, 6, 23, 11, 35, 13, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000), 'EDT')),
 datetime.datetime(2023, 6, 23, 11, 26, 23, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000), 'EDT')),
 datetime.datetime(2023, 6, 23, 11, 55, 33, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000), 'EDT')),
 datetime.datetime(2023, 6, 23, 13, 13, 41, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000), 'EDT')),
 datetime.datetime(2023, 6, 23, 11, 32, 54, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000), 'EDT')),
 datetime.datetime(2023, 6, 23, 13, 17, 4, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000), 'EDT')),
 datetime.datetime(2023, 

In [9]:
df = (
    pd.DataFrame(datetimes, columns=['datetime'])
    .assign(
        color= lambda df_: [fake.color_name() for i in range(len(df_))]
    )
    .sort_values('datetime')
    .reset_index(drop=True)
)
df.head(100)

datetime              color
0  2023-06-23 11:25:16-04:00      LightSeaGreen
1  2023-06-23 11:25:16-04:00            HotPink
2  2023-06-23 11:25:17-04:00          AliceBlue
3  2023-06-23 11:25:17-04:00      DarkTurquoise
4  2023-06-23 11:25:17-04:00          SlateGray
..                       ...                ...
95 2023-06-23 11:26:27-04:00          RoyalBlue
96 2023-06-23 11:26:27-04:00              Azure
97 2023-06-23 11:26:28-04:00  MediumSpringGreen
98 2023-06-23 11:26:28-04:00       LemonChiffon
99 2023-06-23 11:26:29-04:00           DarkBlue

[100 rows x 2 columns]

In [10]:
df.to_csv('/Users/dataders/repos/jaffle_shop/seeds/timecolor.csv', index=False)

In [37]:
df.nunique()

DateTime    5409
color        140
dtype: int64

- seed table `timecolor.csv` 10k rows generated with Faker with two columns
  - `datetime`: timestamps from today ranging from 11 AM to 6 PM
  - `color`: random selection of 140 colors
- view `my_view.sql`
  - meant to mock a streaming source that grows over time
  - definition: `SELECT * FROM {{ ref('timecolor') }} WHERE datetime <= CURRENT_TIMESTAMP()` 
- dynamic table `my_dt.sql`: `select * from my_view` with a  `target_lag: 1 min`


select * from my_mv